slowquant/qiskit_interface/wavefunction.py

```diff
@@ -293,19 +314,38 @@ def change_primitive(self, primitive: BaseEstimator | BaseSampler, verbose: bool
         self.QI.total_device_calls = 0
         self.QI.total_shots_used = 0
         self.QI.total_paulis_evaluated = 0
-        self.QI.cliques = Clique()
+        self.QI._reset_cliques(verbose=verbose)  # pylint: disable=protected-access
         self.QI._Minv = None  # pylint: disable=protected-access
-        self.QI._primitive = primitive  # pylint: disable=protected-access
-        # IMPORTANT: Shot number in primitive gets always overwritten if a shot number is defined in QI!
-        if self.QI.shots is not None and verbose:
-            print(
-                "Number of shots defined in new primitive are ignored as there is a number defined in the QI of ",
-                self.QI.shots,
-            )
-            print("If you want to change the number of shots, do this manually.")
-            print("Set the number of shots manually to None if you run an ideal simulator.")
-        self.QI.shots = self.QI.shots  # Redo shot check with new primitive
-        self.QI.ISA = self.QI.ISA  # Redo ISA parameter check
+
+        # Reset circuit and initiate re-transpiling
+        ISA_old = self.QI.ISA
+        self._reconstruct_circuit()  # Reconstruct circuit but keeping parameters
+        self.QI.ISA = ISA_old  # Redo ISA including transpilation if requested
```

Comment:
What is happening here? xd

Reply:
"doing sampling runs on hardware and we want to ensure to always transpile to the currently best qubits."

What does this mean? If we change qubit, then we need to recalculate M0.
Maybe I am just misunderstanding

Reply:
Yes. change_primitive() also resets M. 
The new M uses the same layout. In the new QI, the circuit is transpiled once and this layout is kept for all tasks, including M. If needed, additional gates are added in the correct layout.